### PR TITLE
Added headlight, neon light and custom tire options to LSC.

### PR DESCRIPTION
**Pull request by aa15032261**
Also added update feature for vehicle data.

### CODE REVIEW

File: BigBaseV2/src/views/vehicle/view_lsc.cpp [hunk 1]
```diff
@@ -176,23 +205,17 @@ namespace big
 
 		if (ImGui::Checkbox("Bulletproof Tires", &can_tires_burst))
 		{
-			g_fiber_pool->queue_job([] {
-				VEHICLE::SET_VEHICLE_TYRES_CAN_BURST(player_vehicle, !can_tires_burst);
-			});
+			VEHICLE::SET_VEHICLE_TYRES_CAN_BURST(player_vehicle, !can_tires_burst);
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_lsc.cpp]: Calling natives outside of script threads is detected.

Use the fiber pool please.

File: BigBaseV2/src/views/vehicle/view_lsc.cpp [hunk 2]
```diff
@@ -176,23 +205,17 @@ namespace big
...
 		}
 		ImGui::SameLine();
 		if (ImGui::Checkbox("Turbo", &turbo))
 		{
-			g_fiber_pool->queue_job([] {
-				VEHICLE::TOGGLE_VEHICLE_MOD(player_vehicle, MOD_TURBO, turbo);
-			});
+			VEHICLE::TOGGLE_VEHICLE_MOD(player_vehicle, MOD_TURBO, turbo);
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_lsc.cpp]: Calling natives outside of script threads is detected.

Use the fiber pool please.

File: BigBaseV2/src/views/vehicle/view_lsc.cpp [hunk 3]
```diff
@@ -176,23 +205,17 @@ namespace big
...
 		}
 		ImGui::SameLine();
 		if (ImGui::Checkbox("Tiresmoke", &tiresmoke))
 		{
-			g_fiber_pool->queue_job([] {
-				VEHICLE::TOGGLE_VEHICLE_MOD(player_vehicle, MOD_TYRE_SMOKE, tiresmoke);
-			});
+			VEHICLE::TOGGLE_VEHICLE_MOD(player_vehicle, MOD_TYRE_SMOKE, tiresmoke);
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_lsc.cpp]: Calling natives outside of script threads is detected.

Use the fiber pool please.

File: BigBaseV2/src/views/vehicle/view_lsc.cpp [hunk 4]
```diff
@@ -221,35 +244,78 @@ namespace big
...
+
+					for (int i = 0; i < wheel_mods.size(); i++)
+					{
+						auto mod = wheel_mods[i];
+
+						std::string label = "Style " + std::to_string(i);
+
+						if (ImGui::Selectable(label.c_str(), mod == wheel_mod))
+						{
+							VEHICLE::SET_VEHICLE_MOD(player_vehicle, selected_slot, mod, true);
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_lsc.cpp]: Calling natives outside of script threads is detected.

Use the fiber pool please.

File: BigBaseV2/src/views/vehicle/view_lsc.cpp [hunk 5]
```diff
@@ -259,37 +325,27 @@ namespace big
 
 		if (ImGui::Checkbox("Headlight##headlight_en", &xenon))
 		{
-			g_fiber_pool->queue_job([] {
-				VEHICLE::TOGGLE_VEHICLE_MOD(player_vehicle, MOD_XENON_LIGHTS, xenon);
-			});
+			VEHICLE::TOGGLE_VEHICLE_MOD(player_vehicle, MOD_XENON_LIGHTS, xenon);
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_lsc.cpp]: Calling natives outside of script threads is detected.

Use the fiber pool please.

File: BigBaseV2/src/views/vehicle/view_lsc.cpp [hunk 6]
```diff
@@ -259,37 +325,27 @@ namespace big
...
 		}
 		ImGui::SameLine();
 		if (ImGui::Checkbox("Left", &neon_left))
 		{
-			g_fiber_pool->queue_job([] {
-				VEHICLE::SET_VEHICLE_NEON_LIGHT_ENABLED_(player_vehicle, NEON_LEFT, neon_left);
-			});
+			VEHICLE::SET_VEHICLE_NEON_LIGHT_ENABLED_(player_vehicle, NEON_LEFT, neon_left);
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_lsc.cpp]: Calling natives outside of script threads is detected.

Use the fiber pool please.

File: BigBaseV2/src/views/vehicle/view_lsc.cpp [hunk 7]
```diff
@@ -259,37 +325,27 @@ namespace big
...
 		}
 		ImGui::SameLine();
 		if (ImGui::Checkbox("Right", &neon_right))
 		{
-			g_fiber_pool->queue_job([] {
-				VEHICLE::SET_VEHICLE_NEON_LIGHT_ENABLED_(player_vehicle, NEON_RIGHT, neon_right);
-			});
+			VEHICLE::SET_VEHICLE_NEON_LIGHT_ENABLED_(player_vehicle, NEON_RIGHT, neon_right);
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_lsc.cpp]: Calling natives outside of script threads is detected.

Use the fiber pool please.

File: BigBaseV2/src/views/vehicle/view_lsc.cpp [hunk 8]
```diff
@@ -259,37 +325,27 @@ namespace big
...
 		}
 		ImGui::SameLine();
 		if (ImGui::Checkbox("Front", &neon_front))
 		{
-			g_fiber_pool->queue_job([] {
-				VEHICLE::SET_VEHICLE_NEON_LIGHT_ENABLED_(player_vehicle, NEON_FRONT, neon_front);
-			});
+			VEHICLE::SET_VEHICLE_NEON_LIGHT_ENABLED_(player_vehicle, NEON_FRONT, neon_front);
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_lsc.cpp]: Calling natives outside of script threads is detected.

Use the fiber pool please.

File: BigBaseV2/src/views/vehicle/view_lsc.cpp [hunk 9]
```diff
@@ -259,37 +325,27 @@ namespace big
...
 		}
 		ImGui::SameLine();
 		if (ImGui::Checkbox("Back", &neon_back))
 		{
-			g_fiber_pool->queue_job([] {
-				VEHICLE::SET_VEHICLE_NEON_LIGHT_ENABLED_(player_vehicle, NEON_BACK, neon_back);
-			});
+			VEHICLE::SET_VEHICLE_NEON_LIGHT_ENABLED_(player_vehicle, NEON_BACK, neon_back);
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_lsc.cpp]: Calling natives outside of script threads is detected.

Use the fiber pool please.

File: BigBaseV2/src/views/vehicle/view_lsc.cpp [hunk 10]
```diff
@@ -478,13 +504,10 @@ namespace big
...
 						if (ImGui::Selectable(name.c_str(), false))
 						{
-							color[0] = (float)rgb[0] / 255;
-							color[1] = (float)rgb[1] / 255;
-							color[2] = (float)rgb[2] / 255;
-
-							g_fiber_pool->queue_job([] {
-								VEHICLE::SET_VEHICLE_TYRE_SMOKE_COLOR(player_vehicle, (int)(color[0] * 255), (int)(color[1] * 255), (int)(color[2] * 255));
-							});
+							VEHICLE::SET_VEHICLE_TYRE_SMOKE_COLOR(player_vehicle, rgb[0], rgb[1], rgb[2]);
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_lsc.cpp]: Calling natives outside of script threads is detected.

Use the fiber pool please.

File: BigBaseV2/src/views/vehicle/view_lsc.cpp [hunk 11]
```diff
@@ -500,13 +523,10 @@ namespace big
...
 						if (ImGui::Selectable(name.c_str(), false))
 						{
-							color[0] = (float)rgb[0] / 255;
-							color[1] = (float)rgb[1] / 255;
-							color[2] = (float)rgb[2] / 255;
-
-							g_fiber_pool->queue_job([] {
-								VEHICLE::SET_VEHICLE_NEON_LIGHTS_COLOUR_(player_vehicle, (int)(color[0] * 255), (int)(color[1] * 255), (int)(color[2] * 255));
-							});
+							VEHICLE::SET_VEHICLE_NEON_LIGHTS_COLOUR_(player_vehicle, rgb[0], rgb[1], rgb[2]);
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_lsc.cpp]: Calling natives outside of script threads is detected.

Use the fiber pool please.

File: BigBaseV2/src/views/vehicle/view_lsc.cpp [hunk 12]
```diff
@@ -518,23 +538,25 @@ namespace big
...
+					VEHICLE::SET_VEHICLE_CUSTOM_PRIMARY_COLOUR(player_vehicle, color_rgb[0], color_rgb[1], color_rgb[2]);
+					break;
+				case 1:
+					VEHICLE::SET_VEHICLE_CUSTOM_SECONDARY_COLOUR(player_vehicle, color_rgb[0], color_rgb[1], color_rgb[2]);
+					break;
+				case 2:
+					VEHICLE::SET_VEHICLE_TYRE_SMOKE_COLOR(player_vehicle, color_rgb[0], color_rgb[1], color_rgb[2]);
+					break;
+				case 3:
+					VEHICLE::SET_VEHICLE_NEON_LIGHTS_COLOUR_(player_vehicle, color_rgb[0], color_rgb[1], color_rgb[2]);
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_lsc.cpp]: Calling natives outside of script threads is detected.

Use the fiber pool please.

File: BigBaseV2/src/views/vehicle/view_lsc.cpp [hunk 13]
```diff
@@ -584,9 +606,7 @@ namespace big
 							secondary_color = COLOR_CHROME;
 						}
 
-						g_fiber_pool->queue_job([] {
-							VEHICLE::SET_VEHICLE_COLOURS(player_vehicle, primary_color, secondary_color);
-						});
+						VEHICLE::SET_VEHICLE_COLOURS(player_vehicle, primary_color, secondary_color);
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_lsc.cpp]: Calling natives outside of script threads is detected.

Use the fiber pool please.

File: BigBaseV2/src/views/vehicle/view_lsc.cpp [hunk 14]
```diff
@@ -607,9 +627,7 @@ namespace big
 								secondary_color = color;
 							}
 
-							g_fiber_pool->queue_job([] {
-								VEHICLE::SET_VEHICLE_COLOURS(player_vehicle, primary_color, secondary_color);
-							});
+							VEHICLE::SET_VEHICLE_COLOURS(player_vehicle, primary_color, secondary_color);
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_lsc.cpp]: Calling natives outside of script threads is detected.

Use the fiber pool please.

File: BigBaseV2/src/views/vehicle/view_lsc.cpp [hunk 15]
```diff
@@ -632,9 +650,7 @@ namespace big
...
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_lsc.cpp]: Calling natives outside of script threads is detected.

Use the fiber pool please.

File: BigBaseV2/src/views/vehicle/view_lsc.cpp [hunk 16]
```diff
@@ -657,9 +673,7 @@ namespace big
...
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_lsc.cpp]: Calling natives outside of script threads is detected.

Use the fiber pool please.

File: BigBaseV2/src/views/vehicle/view_lsc.cpp [hunk 17]
```diff
@@ -674,9 +688,7 @@ namespace big
 							selected_color = color;
 							pearlescent = color;
 
-							g_fiber_pool->queue_job([] {
-								VEHICLE::SET_VEHICLE_EXTRA_COLOURS(player_vehicle, pearlescent, wheel_color);
-							});
+							VEHICLE::SET_VEHICLE_EXTRA_COLOURS(player_vehicle, pearlescent, wheel_color);
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_lsc.cpp]: Calling natives outside of script threads is detected.

Use the fiber pool please.

File: BigBaseV2/src/views/vehicle/view_lsc.cpp [hunk 18]
```diff
@@ -691,9 +703,7 @@ namespace big
 							selected_color = color;
 							wheel_color = color;
 
-							g_fiber_pool->queue_job([] {
-								VEHICLE::SET_VEHICLE_EXTRA_COLOURS(player_vehicle, pearlescent, wheel_color);
-							});
+							VEHICLE::SET_VEHICLE_EXTRA_COLOURS(player_vehicle, pearlescent, wheel_color);
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_lsc.cpp]: Calling natives outside of script threads is detected.

Use the fiber pool please.

File: BigBaseV2/src/views/vehicle/view_lsc.cpp [hunk 19]
```diff
@@ -708,9 +718,7 @@ namespace big
 							selected_color = color;
 							interior_color = color;
 
-							g_fiber_pool->queue_job([] {
-								VEHICLE::SET_VEHICLE_INTERIOR_COLOR_(player_vehicle, interior_color);
-							});
+							VEHICLE::SET_VEHICLE_INTERIOR_COLOR_(player_vehicle, interior_color);
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_lsc.cpp]: Calling natives outside of script threads is detected.

Use the fiber pool please.

File: BigBaseV2/src/views/vehicle/view_lsc.cpp [hunk 20]
```diff
@@ -725,9 +733,7 @@ namespace big
 							selected_color = color;
 							dashboard_color = color;
 
-							g_fiber_pool->queue_job([] {
-								VEHICLE::SET_VEHICLE_DASHBOARD_COLOR_(player_vehicle, dashboard_color);
-							});
+							VEHICLE::SET_VEHICLE_DASHBOARD_COLOR_(player_vehicle, dashboard_color);
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_lsc.cpp]: Calling natives outside of script threads is detected.

Use the fiber pool please.

File: BigBaseV2/src/views/vehicle/view_lsc.cpp [hunk 21]
```diff
@@ -742,9 +748,7 @@ namespace big
 							selected_color = color;
 							headlight_color = color;
 
-							g_fiber_pool->queue_job([] {
-								VEHICLE::SET_VEHICLE_XENON_LIGHTS_COLOR_(player_vehicle, headlight_color);
-							});
+							VEHICLE::SET_VEHICLE_XENON_LIGHTS_COLOR_(player_vehicle, headlight_color);
```
Comment (Yimura) [inline BigBaseV2/src/views/vehicle/view_lsc.cpp]: Calling natives outside of script threads is detected.

Use the fiber pool please.

**Review by Yimura — CHANGES_REQUESTED** [review]

**Review by Yimura — DISMISSED** [review]
Cool idea to use the `ETag` header as a reference point to decide when to update the local resource.

Haven't tested the code myself yet, but all looks good right now.

**Comment by aa15032261** [comment]
New changes:
- Introduced speed units to the program, which makes speed controls easier to use.
- Fixed issue #333 

image

image

**Comment by howdoiusekeyboard** [comment]
Great work! The bugs have been fixed.
Small observations:
1. Vehicle fly speed doesn't apply it don't stop is checked
2. In LSC in color options, Tire smake should be renamed to tire smoke and should be hidden/greyed out (like headlights in color option is greyed out if Headlight is not enabled in neon light options) if tiresmoke is not enabled in mod options. Because doing otherwise crashes game

**Comment by aa15032261** [comment]
> Great work! The bugs have been fixed. Small observations:
> 
> 1. Vehicle fly speed doesn't apply it don't stop is checked
> 2. In LSC in color options, Tire smake should be renamed to tire smoke and should be hidden/greyed out (like headlights in color option is greyed out if Headlight is not enabled in neon light options) if tiresmoke is not enabled in mod options. Because doing otherwise crashes game

I've fixed item 2 in the latest commit. For item 1, when don't stop is enabled, the program applies the speed as force to the vehicle so the vehicle immediately reaches the max speed. I'm not sure if this behavior is intentional as I'm not the creator of this feature.

**Comment by howdoiusekeyboard** [comment]
Yes, it has been fixed and works as intended
I found some bugs in LSC
1. 
image
2. 
image
image

**Comment by aa15032261** [comment]
> Yes, it has been fixed and works as intended I found some bugs in LSC

Could you tell me how to reproduce these bugs?

**Comment by howdoiusekeyboard** [comment]
> > Yes, it has been fixed and works as intended I found some bugs in LSC
> 
> Could you tell me how to reproduce these bugs?

Are you on discord? This is my id 
I can stream and show it to you there

**Comment by aa15032261** [comment]
> > > Yes, it has been fixed and works as intended I found some bugs in LSC
> > 
> > 
> > Could you tell me how to reproduce these bugs?
> 
> Are you on discord? This is my id 尺ㄚᗪ乇尺爪ㄖ几Ҝ丨丨丨#9089 I can stream and show it to you there

Nvm, I have figured it out and it is fixed in the latest commit.

**Comment by howdoiusekeyboard** [comment]
> > > > Yes, it has been fixed and works as intended I found some bugs in LSC
> > > 
> > > 
> > > Could you tell me how to reproduce these bugs?
> > 
> > 
> > Are you on discord? This is my id 尺ㄚᗪ乇尺爪ㄖ几Ҝ丨丨丨#9089 I can stream and show it to you there
> 
> Nvm, I have figured it out and it is fixed in the latest commit.

Yes, a 3rd empty box bug with some mod options has been fixed and is much better but,
Front wheel option doesn't work and doesn't show the correct wheel (also didn't understand the naming)
1. I choose offroad in wheel type and it shows high end wheels set in front wheel and they don't apply, plus game crash

image


{Misc: Teleport to PV should be renamed to Teleport in PV}

**Comment by aa15032261** [comment]
> > > > > Yes, it has been fixed and works as intended I found some bugs in LSC
> > > > 
> > > > 
> > > > Could you tell me how to reproduce these bugs?
> > > 
> > > 
> > > Are you on discord? This is my id 尺ㄚᗪ乇尺爪ㄖ几Ҝ丨丨丨#9089 I can stream and show it to you there
> > 
> > 
> > Nvm, I have figured it out and it is fixed in the latest commit.
> 
> Yes, a 3rd empty box bug with some mod options has been fixed and is much better but, Front wheel option doesn't work and doesn't show the correct wheel (also didn't understand the naming)
> 
> 1. I choose offroad in wheel type and it shows high end wheels set in front wheel and they don't apply, plus game crash
> 
> image
> 
> {Misc: Teleport to PV should be renamed to Teleport in PV}



> 


The issue should be fixed in the latest commit.

**Comment by Yimura** [comment]
Tell me when  you're done with, though don't add any more features to this PR.

PR's are supposed to be a single feature or bug fix, if you want to fix bugs and do other things then you're supposed to make a PR for each.

**Comment by aa15032261** [comment]
> Tell me when you're done with, though don't add any more features to this PR.
> 
> PR's are supposed to be a single feature or bug fix, if you want to fix bugs and do other things then you're supposed to make a PR for each.

I'll keep that in mind, thanks! And yes, the code is now finalized.

**Comment by howdoiusekeyboard** [comment]
Amazing work! I have tested thoroughly and I found no bugs

**Review by Yimura — APPROVED** [review]

**Pull request merged**